apex/contrib/optimizers/distributed_fused_adam.py

```diff
@@ -648,95 +647,152 @@ def grad_sync(self):
                     False,
                 )[0] ** 2
             else:
-                non_parallel_grad_norm_sq = torch.zeros([1], device=self.device)
+                grad_norm_sq = torch.zeros([1], dtype=torch.float32, device=self.device)
+
+        return grad_norm_sq.detach().view([])
+
+    def grad_norm(self, parameters=[], norm_type=2.0, force=False):
+        """Gradient norm of parameters in optimizer
+
+        The norm is computed over all gradients together, as if they
+        were concatenated into a single vector. All provided
+        parameters must be managed by optimizer.
+
+        The computed value is cached to avoid redundant communication.
+
+        Arguments:
+            parameters (iterable, optional): an iterable of parameters
+                in optimizer (default: all parameters in optimizer).
+            norm_type (float or int, optional): type of the used
+                p-norm (default: 2). Only 2-norm is currently
+                supported.
+            force (bool, optional): ignore cached value and force norm
+                computation (default: False).
+
+        """
+        if force or self._grad_norm is None:
+            norm_type = float(norm_type)
+            assert norm_type == 2.0
+            grad_norm_sq = self._local_grad_norm(
+                parameters=parameters,
+                norm_type=norm_type,
+            )
             torch.distributed.all_reduce(
-                non_parallel_grad_norm_sq,
+                grad_norm_sq,
+                op=torch.distributed.ReduceOp.SUM,
                 group=self.distributed_process_group,
             )
-            grad_norm_sq -= non_parallel_grad_norm_sq
+            self._grad_norm = grad_norm_sq.sqrt()
+        return self._grad_norm.detach()
 
-        return grad_norm_sq.sqrt()
+    def clip_grad_norm(self, max_norm, parameters=[], norm_type=2.0):
+        """Clips gradient norm of parameters in optimizer
 
-    def step(self, closure=None, scale=1.):
+        The norm is computed over all gradients together, as if they
+        were concatenated into a single vector. The scaling is
+        deferred until the optimizer step, which should be called
+        immediately after this function.
+
+        The computed grad norm is cached to avoid redundant
+        communication.
+
+        Arguments:
+            max_norm (float or int): max norm of the gradients
+            parameters (iterable, optional): an iterable of parameters
+                in optimizer (default: all parameters in optimizer).
+            norm_type (float or int, optional): type of the used
+                p-norm (default: 2)
+
+        """
+        assert max_norm > 0
+        total_norm = self.grad_norm(parameters=parameters, norm_type=norm_type)
+        inv_clip_coef = (total_norm + 1e-6) / max_norm
+        self._inv_grad_scale = torch.clamp(inv_clip_coef, min=1.0).view(1)
+        return total_norm
+
+    def step(self, grad_scaler=None):
```

Comment:
also, you can define `grad_scaler` as a keyword only argument given https://github.com/pytorch/pytorch/blob/fe73528a904aa0627b5a6f7376a2331cb5985bfa/torch/cuda/amp/grad_scaler.py#L329.